src/NServiceBus.Core/Routing/SpecificInstanceDistributionPolicy.cs

```diff
@@ -0,0 +1,42 @@
+namespace NServiceBus
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+    using Routing;
+
+    class SpecificInstanceDistributionPolicy : IDistributionPolicy
+    {
+        readonly string specificInstance;
+
+        public SpecificInstanceDistributionPolicy(string specificInstance)
+        {
+            this.specificInstance = specificInstance;
+        }
+
+        public DistributionStrategy GetDistributionStrategy(string endpointName)
+        {
+            return new SpecificInstanceDistributionStrategy(specificInstance);
+        }
+
+        class SpecificInstanceDistributionStrategy : DistributionStrategy
+        {
+            public SpecificInstanceDistributionStrategy(string specificInstance)
+            {
+                this.specificInstance = specificInstance;
+            }
+
+            public override IEnumerable<UnicastRoutingTarget> SelectDestination(IList<UnicastRoutingTarget> allInstances)
+            {
+                var target = allInstances.FirstOrDefault(t => t.Instance != null && t.Instance.Discriminator == specificInstance);
+                if (target == null)
+                {
+                    throw new Exception($"Specified instance {specificInstance} has not been configured in the routing tables.");
+                }
+                yield return target;
```

Comment:
@danielmarbach I did spend some time comparing arrays vs. yielding using BenchmarkDotNet:

```ini

BenchmarkDotNet=v0.9.7.0
OS=Microsoft Windows NT 6.2.9200.0
Processor=Intel(R) Core(TM) i7-2600K CPU 3.40GHz, ProcessorCount=8
Frequency=3331177 ticks, Resolution=300.1942 ns, Timer=TSC
HostCLR=MS.NET 4.0.30319.42000, Arch=64-bit RELEASE [RyuJIT]
JitModules=clrjit-v4.6.1080.0

Type=YieldTests  Mode=Throughput  

```
                Method | NumberOfEnumerations |      Median |     StdDev |    Gen 0 | Gen 1 | Gen 2 | Bytes Allocated/Op |
---------------------- |--------------------- |------------ |----------- |--------- |------ |------ |------------------- |
              **UseYield** |                    **1** |  **28.3400 ns** |  **0.4149 ns** |   **251.62** |     **-** |     **-** |              **22.50** |
 UseArrayAsIEnumerable |                    1 |  36.0879 ns |  0.3318 ns |   427.12 |     - |     - |              38.17 |
       UseArrayAsArray |                    1 |  16.6967 ns |  0.2548 ns |   270.70 |     - |     - |              24.28 |
              **UseYield** |                    **5** | **144.6804 ns** |  **1.3118 ns** | **1,376.72** |     **-** |     **-** |             **123.03** |
 UseArrayAsIEnumerable |                    5 | 146.8555 ns |  2.6982 ns |   928.72 |     - |     - |              83.10 |
       UseArrayAsArray |                    5 |  20.0769 ns |  0.4160 ns |   242.19 |     - |     - |              21.73 |
              **UseYield** |                   **10** | **341.6629 ns** | **44.8241 ns** | **3,014.23** |     **-** |     **-** |             **269.18** |
 UseArrayAsIEnumerable |                   10 | 296.8246 ns | 11.5910 ns | 1,781.00 |     - |     - |             159.48 |
       UseArrayAsArray |                   10 |  26.4914 ns |  0.7516 ns |   240.92 |     - |     - |              21.54 |


Clearly, returning an array and then enumerating it as an array is the best performing, regardless of how many times you enumerate it.

For the options that don't require us to alter the return type (or make an assumption and cast), the it depends on the number of times the result will be enumerated. Since this code only enumerates this collection once, based on these results, sticking with `yield` seems like the best choice.